### PR TITLE
3.x

### PR DESCRIPTION
**Pull request by anowell**
Changelog captures the majority of what's going on here:

**Breaking changes**
- Combined `AlgoInput` and `AlgoOutput` into a simpler `AlgoIo` that removes references that weren't being used.
- Removed `algo::version` module. Use `user/algo/version` string instead.
- Moved entrypoint traits into new root module.
- Moved entrypoint functionality behind `entrypoint` feature flag.
- Replace `algo_entrypoint!` macro with `#[entrypoint]` attribute.
- Algorithmia client instantiation is now fallible instead of hiding errors until HTTP request
- `ApiAuth` type has been made private
- Entrypoint traits accept `&mut self` to allow for easier state manipulation
- Entrypoint codegen autoboxes return types (for lack of specialization)

### CODE REVIEW

**Comment by anowell** [comment]
Maybe the easiest way to see the differences is:
https://docs.rs/algorithmia/2.1.3/algorithmia/
vs 
https://docs.rs/algorithmia/3.0.0-beta/algorithmia/

File: algorithmia/README.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+[Documentation](http://algorithmiaio.github.io/algorithmia-rust/algorithmia/)
+
+[![Build Status](https://travis-ci.org/algorithmiaio/algorithmia-rust.svg)](https://travis-ci.org/algorithmiaio/algorithmia-rust)
+[![Crates.io](https://img.shields.io/crates/v/algorithmia.svg?maxAge=2592000)](https://crates.io/crates/algorithmia)
+
+## Getting started
+
+The Algorithmia Rust client is published to [crates.io](https://crates.io/crates/algorithmia).
+Add `algorithmia = "2.0.0"` to the dependencies in your `Cargo.toml` and run `cargo install`.
```
Comment (zeryx) [inline algorithmia/README.md]: 3.0.0 right?

File: algorithmia/src/data/file.rs [hunk 2]
```diff
@@ -68,128 +91,115 @@ impl DataFile {
...
+            .put(url)
+            .body(body)
+            .send()
             .chain_err(|| ErrorKind::Http(format!("writing file '{}'", self.to_data_uri())))?;
         let mut res_json = String::new();
         res.read_to_string(&mut res_json)
             .chain_err(|| ErrorKind::Io(format!("writing file '{}'", self.to_data_uri())))?;
 
-        match *res.status() {
+        match res.status() {
```
Comment (zeryx) [inline algorithmia/src/data/file.rs]: this is so much cleaner, wow

**Review by zeryx — COMMENTED** [review]
Initial code change investigations all make sense!

**Review by zeryx — APPROVED** [review]
Seems solid to me, one nit I'd have (we spoke about this already) is to add some documentation describing the `Result<T>` type that hides the 'E' of the default `Result<T,E>` type. Making this change could improve error handling comprehension on how to use this package 

I can't see any reason why that should be a blocker though, the rest of it is :ship: able.

File: algorithmia/README.md (hunk 1 above)
Comment (anowell) [reply algorithmia/README.md]: good catch

**Comment by anowell** [comment]
This also breaks compat because of serde upgrade (serde 1.0) and by completely removing support for `rustc_serialize`.

**Comment by anowell** [comment]
Switched to `Result<T, E>` long form in https://github.com/algorithmiaio/algorithmia-rust/pull/8/commits/4f0e388a3d796e5e0d1f231e661a1818f9b11538

**Comment by anowell** [comment]
I'm not yet happy with how error chaining works, and I'd rather not depend on `error_chain` any more for this crate, so I'm refactoring that, which will hopefully be the last big changes before cutting beta2

**Comment by zeryx** [comment]
> This also breaks compat because of serde upgrade (serde 1.0) and by completely removing support for `rustc_serialize`.

As we've already mentioned the breaking changes, and the fact that IPA won't be BWC anyways I think this is totally fine

**Pull request merged**